akka-distributed-data/src/main/scala/akka/cluster/ddata/GCounter.scala

```diff
@@ -39,12 +39,14 @@ object GCounter {
  */
 @SerialVersionUID(1L)
 final class GCounter private[akka] (
-  private[akka] val state: Map[UniqueAddress, BigInt] = Map.empty)
-  extends ReplicatedData with ReplicatedDataSerialization with RemovedNodePruning with FastMerge {
+  private[akka] val state: Map[UniqueAddress, BigInt] = Map.empty,
+  override val delta:      Option[GCounter]           = None)
```

Comment:
In the end we need both the delta and the full state for an update operation. I was first thinking of returning a tuple, but that doesn't compose well. For example I like api:

```scala
GSet.empty[String].add("a").add("b")
```

The mutation could return the delta instead of the state, and the state could be derived by merging the delta into the previous state. I see a few disadvantages with that:

* It would be different api (behavior) for delta-CRDT and full-state-only-CRDT. Rather confusing. I also find it more natural that a mutator actually return the full state from a user api perspective.
* Merging the delta back would be less efficient, and that would be noticeable. That's why I previously did merge optimizations with the `ancestor` field in some types.

Therefore I suggest that the mutators return the full state, in the same way as old implementations, and that the delta is collected "on the side" and accessed via the `delta` method by the `Replicator`.

> You cannot have multiple deltas with your current state. Ideally, the delta is the exact difference between two replicas. Depending on the protocol (many possibilities here), the difference may be replica-pair-specific. This choice therefore seems to eliminate possible protocol implementations.

Such deltas are delta-groups or delta-intervals (merging of individual deltas to create new delta). That is something the Replicator can do. The individual delta must be created by the mutation operation.

> Each delta-CRDT mutation must effectively implement merge, or fall back to computing the delta and then calling merge, which is not DRY.

I'm not sure I see a problem with this. The best knowledge to most efficiently create next full state and delta is in the mutation, e.g. `GSet.add`.